Sources/GRPC/GRPCClientStateMachine.swift

```diff
@@ -0,0 +1,588 @@
+  /// - Parameter allocator: A `ByteBufferAllocator` to allocate the buffer into which the encoded
+  ///     request will be written.
+  mutating func sendRequest(
+    _ message: Request,
+    allocator: ByteBufferAllocator
+  ) -> Result<ByteBuffer, MessageWriteError> {
+    return self.state.sendRequest(message, allocator: allocator)
+  }
+
+  /// Terminates the request stream.
+  ///
+  /// The client must be streaming requests in order to terminate the request stream. Valid
+  /// states transitions are:
+  /// - `.clientStreamingServerIdle` → `.clientClosedServerIdle`
+  /// - `.clientStreamingServerStreaming` → `.clientClosedServerStreaming`
+  ///
+  /// It is invalid (but not fatal) for the client to attempt to close the request stream multiple
+  /// times The states in which this is possible are:
+  /// - `.clientClosedServerIdle`
+  /// - `.clientClosedServerStreaming`
+  /// - `.clientClosedServerClosed`
+  ///
+  /// Closing the request stream is invald, and fatal, for the following states:
+  /// - `.clientIdleServerIdle`
+  mutating func sendEndOfRequestStream() -> Result<Void, SendEndOfRequestStreamError> {
+    return self.state.sendEndOfRequestStream()
+  }
+
+  /// Recieve an acknowledgement of the RPC from the server. This **must not** be a "trailers-only"
+  /// response.
+  ///
+  /// The server must be idle in order to recive response headers. The valid state transitions are:
+  /// - `.clientStreamingServerIdle` → `.clientStreamingServerStreaming`
+  /// - `.clientClosedServerIdle` → `.clientClosedServerStreaming`
+  ///
+  /// It is invalid and fatal for the RPC to recieve response headers from the following states:
+  /// - `.clientIdleServerIdle`
+  /// - `.clientStreamingServerStreaming`
+  /// - `.clientClosedServerStreaming`
+  /// - `.clientClosedServerClosed`
+  ///
+  /// - Parameter headers: The headers recieved from the server.
+  mutating func recieveResponseHeaders(
+    _ responseHead: HTTPResponseHead
+  ) -> Result<HTTPHeaders, RecieveResponseHeadError> {
+    return self.state.recieveResponseHeaders(responseHead, logger: self.logger)
+  }
+
+  /// Read a response buffer from the server and return any decoded messages.
+  ///
+  /// If the response stream has an expected count of `.one` then this function is guaranteed to
+  /// produce *at most* one `Response` in the `Result`.
+  ///
+  /// To recieve a response buffer the server must be streaming. Valid states are:
+  /// - `.clientClosedServerStreaming` → `.clientClosedServerStreaming`
+  /// - `.clientStreamingServerStreaming` → `.clientStreamingServerStreaming`
+  ///
+  /// It is invalid and fatal to recieve a response in the following states:
+  /// - `.clientIdleServerIdle`
+  /// - `.clientClosedServerStreaming`
+  /// - `.clientStreamingServerStreaming`
+  /// - `.clientClosedServerClosed`
+  ///
+  /// - Parameter buffer: A buffer of bytes received from the server.
+  mutating func recieveResponse(
```

Comment:
Calls this "receiveResponse**s**" instead?

Reply:
I think `recieveResponseBuffer` is probably more accurate still.